### PR TITLE
Add jmp_esp/jmp_rsp attribute to ROP

### LINKED ISSUE #1528: ROP: Should have a ROP.jmp_esp / ROP.jmp_rsp gadget
This is an easy gadget for us to detect, and helps out a lot with older challenges that have NX disabled.

### PR DESCRIPTION
**Pull request by saullocarvalho**
Fixes #1528 

I've used `setattr` to set a different attribute name based on the architecture context.
Is there any better way to do this?

I've thought about using `jmp_sp` instead of `jmp_esp`/`jmp_rsp`, but I've decided to solve as the issue title suggests.

### CODE REVIEW

**Comment by saullocarvalho** [comment]
After reading more the ROP source code, I figured out a way of implement ROP.jmp_esp / ROP.jmp_rsp using the `__getattr___` method.

File: pwnlib/rop/rop.py [hunk 1]
```diff
@@ -88,6 +88,17 @@
...
+
+    >>> context.clear(arch='i386')
+    >>> elf = ELF.from_assembly('nop; pop eax; jmp esp; int 0x80; jmp esp; ret')
+    >>> rop = ROP(elf, badchars=b'\x02')
+    >>> jmp_gadget = rop.jmp_esp
+    >>> elf.read(jmp_gadget.address, 2) == asm('jmp esp')
+    True
+    >>> rop = ROP(elf, badchars=b'\x02\x06')
+    >>> rop.jmp_esp == None
+    True
```
Comment (heapcrash) [inline pwnlib/rop/rop.py:120]: Can we add an `amd64` test and separate the 'positive' and 'negative' tests a bit visually?

File: pwnlib/rop/rop.py [hunk 2]
```diff
@@ -1227,6 +1239,23 @@ def __getattr__(self, k):
             leave = None
         self.leave = leave
 
+        #
+        # Load ROP.jmp_esp ('i386') or ROP.jmp_rsp ('amd64') gadgets
+        #
+        jmp_sp = {
+            4: 'jmp esp',
+            8: 'jmp rsp'
+        }[context.bytes]
```
Comment (heapcrash) [inline pwnlib/rop/rop.py]: Don't shell out to `asm()` here, hard-code the byte values since they are known.  `as` is slow and these will never change.  I'd also prefer using `context.arch` instead of `context.bytes`.

File: pwnlib/rop/rop.py [hunk 3]
```diff
@@ -1341,6 +1370,19 @@ def __getattr__(self, attr):
                 count = int(attr.split('_')[1])
             return self.search(move=count)
 
+        #
+        # Check for 'jmp_esp' or 'jmp_rsp'
+        #
+        sp_regs = ['esp', 'rsp']
+
+        if attr.startswith('jmp') \
+        and attr.split('_')[1] in sp_regs:
```
Comment (heapcrash) [inline pwnlib/rop/rop.py]: This seems really error prone, we should just use `ELF.search(..., executable=True)` for the specific byte sequence -- especially since it doesn't change between architectures.

File: pwnlib/rop/rop.py (hunk 2 above)
Comment (heapcrash) [reply pwnlib/rop/rop.py]: On second thought, they're the same for both architectures so we can just hard-code a single value.

**Review by heapcrash — CHANGES_REQUESTED** [review]

File: pwnlib/rop/rop.py (hunk 2 above)
Comment (saullocarvalho) [reply pwnlib/rop/rop.py]: Using the hard-code value is really better. I'm fixing it.

I've used `context.bytes` as it is used at
https://github.com/Gallopsled/pwntools/blob/dev/pwnlib/rop/rop.py#L1183-L1186
. I am only pointing out, because maybe `context.arch` should be used there too.

File: pwnlib/rop/rop.py (hunk 3 above)
Comment (saullocarvalho) [reply pwnlib/rop/rop.py]: My PR loads all `ROP.jmp_$sp` addresses as gadgets at `__load`, filtering out based on `self._badchars`.
If you think it is better, I move the code to `__getattr__`. It will only return the first match that does not have any badchar. Using this approach, I do not need to touch `self.gadgets` while loading.

Comment (heapcrash) [reply pwnlib/rop/rop.py]: I'd rather not filter out JMP gadgets and just search for the specific byte sequence, since this is special-case code anyway.

Maybe in a future PR we can search for all `jmp reg` instances and put them in e.g. `ROP.jmp_gadgets`

For now, let's hard-code the bytes for `jmp $sp` and search with `ELF.search`

File: pwnlib/rop/rop.py (hunk 1 above)
Comment (saullocarvalho) [reply pwnlib/rop/rop.py:120]: I have added the `amd64` test and separated the 'positive' and 'negative' tests.

File: pwnlib/rop/rop.py (hunk 3 above)
Comment (saullocarvalho) [reply pwnlib/rop/rop.py]: I've moved the code to `__getattr__`, used the hard-coded opcodes to search with `ELF.search` until the gadget address does not have any badchar.

**Comment by zachriggle** [comment]
Thanks for the continued work on this, but I'd prefer to just look up the jmp_esp / rsp gadgets one time when we find the rest of the gadgets, instead of every time in __getattr__.  

Using __getattr__ also hides the attribute from autocomplete in e.g. iPython or dir().

Also thanks for adding to the changelog!  Glad that check is having the intended effect.

**Comment by zachriggle** [comment]
Deleted my last comment, apparently I forgot how the code looked in the first place.
